src/signal_fmt_plug.c

```diff
@@ -0,0 +1,294 @@
+	cracked_count = self->params.max_keys_per_crypt;
+}
+
+static void done(void)
+{
+	MEM_FREE(cracked);
+	MEM_FREE(saved_key);
+	MEM_FREE(saved_len);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *p = ciphertext, *ctcopy, *keeptr;
+	int extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LENGTH))
+		return 0;
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "$")) == NULL) // version
+		goto bail;
+	if (!isdec(p))
+		goto bail;
+	if (atoi(p) != 1)
+		goto bail;
+	if ((p = strtokm(NULL, "$")) == NULL) // iterations
+		goto bail;
+	if (!isdec(p))
+		goto bail;
+	if ((p = strtokm(NULL, "$")) == NULL) // encrypted_salt
+		goto bail;
+	if (hexlenl(p, &extra) > 32 * 2 || extra)
+		goto bail;
+	if (!ishexlc(p))
+		goto bail;
+	if ((p = strtokm(NULL, "$")) == NULL) // mac_salt
+		goto bail;
+	if (hexlenl(p, &extra) > 32 * 2 || extra)
+		goto bail;
+	if (!ishexlc(p))
+		goto bail;
+	if ((p = strtokm(NULL, "$")) == NULL) // master_secret
+		goto bail;
+	if (hexlenl(p, &extra) > 128 * 2 || extra)
+		goto bail;
+	if (!ishexlc(p))
+		goto bail;
+	if ((p = strtokm(NULL, "$")) == NULL) // mac
+		goto bail;
+	if (hexlenl(p, &extra) != 20 * 2 || extra)
+		goto bail;
+	if (!ishexlc(p))
+		goto bail;
+
+	MEM_FREE(keeptr);
+	return 1;
+
+bail:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs;
+	int i;
+	char *p = ciphertext, *ctcopy, *keeptr;
+
+	memset(&cs, 0, sizeof(cs));
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	p = strtokm(ctcopy, "$");
+	p = strtokm(NULL, "$");
+	cs.iterations = atoi(p);
+	p = strtokm(NULL, "$");
+	p = strtokm(NULL, "$");
+	cs.mac_salt_size = strlen(p) / 2;
+	for (i = 0; i < cs.mac_salt_size; i++)
+		cs.mac_salt[i] = (atoi16[ARCH_INDEX(p[2*i])] << 4) | atoi16[ARCH_INDEX(p[2*i+1])];
+	p = strtokm(NULL, "$");
+	cs.master_secret_size = strlen(p) / 2;
+	for (i = 0; i < cs.master_secret_size; i++)
+		cs.master_secret[i] = (atoi16[ARCH_INDEX(p[2*i])] << 4) | atoi16[ARCH_INDEX(p[2*i+1])];
+
+	MEM_FREE(keeptr);
+
+	return (void *)&cs;
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index;
+
+	memset(cracked, 0, sizeof(cracked[0]) * cracked_count);
+
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index++) {
+		unsigned char key[16];
+		int keylen = 16;
+		unsigned char hash[20];
+
+		pkcs12_pbe_derive_key(1, cur_salt->iterations,
+				MBEDTLS_PKCS12_DERIVE_KEY,
+				(unsigned char*)saved_key[index],
+				saved_len[index], cur_salt->mac_salt,
+				cur_salt->mac_salt_size, key, keylen);
+		hmac_sha1(key, keylen, cur_salt->master_secret, cur_salt->master_secret_size - 20, hash, 20);
+		cracked[index] = (0 == memcmp(hash, cur_salt->master_secret + cur_salt->master_secret_size - 20, 20));
```

Comment:
We hates this asswards constant-first notation :-)

Either
```c
cracked[index] = (memcmp(hash, cur_salt->master_secret + cur_salt->master_secret_size - 20, 20) == 0);
```

or (this is how I'd write it, fwiw)
```c
cracked[index] = !memcmp(hash, cur_salt->master_secret + cur_salt->master_secret_size - 20, 20);
```

Reply:
Ack, fixing :-)